keps/sig-cli/1442-kubectl-remove-noOptDefVal/README.md

```diff
@@ -0,0 +1,156 @@
+
+An example usage:
+
+```
+kubectl config set-cluster e2e --insecure-skip-tls-verify
+```
+
+In this case, `insecure-skip-tls-verify` is a boolean [flag][k-set-cluster-cmd]. Even though, its value is not explicitly provided, it is assumed to be true since the flag is set. 
+
+Without enabling this option, the command accepts input in two formats: `<flag_name>=<value> `and `<flag_name> <value>.` However, with this option enabled, only the `<flag_name>=<value>` format is accepted. For instance:
+
+```
+kubectl create deploy --image nginx test
+```
+
+creates a deployment with image `nginx` and name `test`, whereas:
+
+```
+kubectl create deploy --image nginx --validate strict
+```
+
+creates a deployment with image `nginx` and name `strict`. 
+
+A few other examples of misintepretation as shared in issue [#1442][issue-1442] by @salavessa are:
+
+1. `kubectl delete pod my-sts --cascade orphan --dry-run server`
+
+Expected: Delete pod with the name `my-sts` with cascade set to `orphan` and `dry-run` to server.
+
+Actual: Tries deleting 3 pods: `my-sts`, `orphan` and `server`.
+```
+W0922 15:21:03.455729   68481 helpers.go:663] --dry-run is deprecated and can be replaced with --dry-run=client.
+pod "my-sts" deleted (dry run)
+pod "orphan" deleted (dry run)
+pod "server" deleted (dry run)
+```
+
+2. `kubectl delete pods my-sts --cascade orphan --dry-run=server`
+
+Expected: Delete pod with the name `my-sts` with cascade set to `orphan` and `dry-run` to server.
+
+Actual: Tries deleting 2 pods: `my-sts` and `orphan`.
+
+```
+pod "my-sts" deleted (dry run)
+Error from server (NotFound): pod "orphan" not found
+```
+
+Removing the use `NoOptDefVal` will help us in providing consistency across the use flags of all types. With this option removed, the user experience would look like:
+
+1. `kubectl delete pod my-sts --cascade orphan --dry-run server`: Delete pod with the name `my-sts` with cascade set to `orphan` and `dry-run` to server.
+2. `kubectl delete pod my-sts --cascade=orphan --dry-run=server`: Delete pod with the name `my-sts` with cascade set to `orphan` and `dry-run` to server.
+3. `kubectl delete pods my-sts --cascade --dry-run=server`: Would error stating `--cascade` flag needs a value.
+4. `kubectl delete pods my-sts --cascade --dry-run`: Would error stating `--cascade` and `--dry-run` flags need values.
+
+Though this would enforce users to specify flags as key value pairs, it would ensure that the command behaves the same whether or not `=` or `space` is provided between the flags and their values.
+
+
+### Risks and Mitigations
+
+This may break existing users that rely on the feature of `NoOptDefVal` and specify just the flags assuming that their value would be set to `true`.
+
+## Design Details
+
+In order to implement this, the use of `NoOptDefVal` needs to be removed from the following places.
```

Comment:
I think, most important part of this KEP is how this change will be managed without breaking existing users and propose an elegant way for smooth switch.

Reply:
@ardaguclu I've updated the KEP to include the deprecation process. Since this is breaking, the best I can think of is having warnings, followed by the removal of the use of `NoOptDefVal` (ie) we error when any flag is provided without a value. I'm not sure on the usual deprecation period which we follow. Are there any other steps I am missing in the process?

We do have a PR that adds linters, to make sure we don't introduce this behaviour again in our codebase: https://github.com/kubernetes/kubernetes/pull/119379

Reply:
But we can not break current users. I think, this KEP should provide a way for smooth transition. For example, for a couple of releases, we can just warn users about the deprecation of these flags and only after that we can start removing.